networking/p2p/src/main/java/tech/pegasys/artemis/networking/p2p/jvmlibp2p/PeerManager.java

```diff
@@ -0,0 +1,137 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.networking.p2p.jvmlibp2p;
+
+import com.google.common.primitives.UnsignedLong;
+import io.libp2p.core.Connection;
+import io.libp2p.core.ConnectionHandler;
+import io.libp2p.core.multiformats.Multiaddr;
+import io.libp2p.network.NetworkImpl;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.Level;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.jetbrains.annotations.NotNull;
+import tech.pegasys.artemis.datastructures.networking.libp2p.rpc.HelloMessage;
+import tech.pegasys.artemis.util.SSZTypes.Bytes4;
+import tech.pegasys.artemis.util.alogger.ALogger;
+
+public class PeerManager implements ConnectionHandler {
+  private static final ALogger STDOUT = new ALogger("stdout");
+  private final ALogger LOG = new ALogger(PeerManager.class.getName());
+
+  private final ScheduledExecutorService scheduler;
+  private static final Duration RECONNECT_TIMEOUT = Duration.ofSeconds(1);
+
+  private ConcurrentHashMap<Multiaddr, Peer> connectedPeerMap = new ConcurrentHashMap<>();
+
+  private final RPCMessageHandler<HelloMessage, HelloMessage> rpcMessageHandler;
+
+  public PeerManager(final ScheduledExecutorService scheduler) {
+    this.scheduler = scheduler;
+
+    this.rpcMessageHandler =
+        new RPCMessageHandler<>("/eth2/beacon_chain/req/hello/1/ssz", this::hello) {
+          @Override
+          protected CompletableFuture<HelloMessage> invokeLocal(
+              Connection connection, HelloMessage helloMessage) {
+            return CompletableFuture.completedFuture(
+                this.helloHandler.apply(connection, helloMessage));
+          }
+        };
+  }
+
+  @Override
+  public void handleConnection(@NotNull final Connection connection) {
+    STDOUT.log(Level.INFO, "New connection: " + connection.remoteAddress());
+    Peer peer = new Peer(connection);
+    connectedPeerMap.put(peer.getPeerMultiaddr(), peer);
+
+    connection
+        .closeFuture()
+        .thenRun(() -> STDOUT.log(Level.INFO, "Peer disconnected: " + peer.getPeerId()));
+
+    if (connection.isInitiator()) {
+      rpcMessageHandler
+          .invokeRemote(connection, createHelloMessage())
+          .thenApply(resp -> peer.getRemoteHelloMessage().complete(resp));
+    }
+  }
+
+  public CompletableFuture<?> connect(final Multiaddr peer, final NetworkImpl network) {
+    STDOUT.log(Level.INFO, "Connecting to " + peer);
+    return network
+        .connect(peer)
+        .whenComplete(
+            (conn, t) -> {
```

Comment:
I know this is tiny, but "t" feels weird for an error response variable. no?

Reply:
agreed.  it was adapted from harmony.  i will modify it to match our style

Reply:
fyi, t is short for Throwable so it isn't terrible, but i don't mind being more explicit with names